TabRuns

claudiaareneee opened the pull request:
## Description

This PR contains the following changes:
* Creates a new type of run called `TabRun`
* Associates the tab styling within `TextStyleSettings`
* Adds tests for new run type
* Adds a new geometry feature for debugging the ranges of individual text runs
* Adds a `isWhitespace` method

## Link to Issue

closes #7678 

## Screenshots

![tabbing](https://github.com/user-attachments/assets/478fba44-76cf-4c93-a46b-aa729d0d5029)

claudiaareneee commented on core/common/src/annotation/TextBlock.ts:
```diff
@@ -497,7 +559,9 @@ export class TextBlock extends TextBlockComponent {
 
-  /** Returns true if every paragraph in this text block is empty. */
+  /** Returns true if every paragraph in this text block is empty.
+   * idk, should I get rid of this? I can't see a use for it. using isWhitespace is probably better in most cases.
+  */
```
I need to remove this comment, but the question is still valid. Is isEmpty useful? I mean it corresponds to createEmpty, but it shouldn't be used when determining whether or not geometry can be created. That would rely on `isWhitespace`

claudiaareneee replied:
Okay, I can leave it then and remove the comment

pmconne commented:
Your screenshot doesn't have any examples of consecutive tab stops - did you test those?
Did you test lines containing only tab stops? Lines ending in tab stops?
Did you test how tab stops behave with line wrapping?

pmconne reviewed (approved):
Just need a bunch more tests especially around edges cases.

claudiaareneee commented:
@pmconne, I added tests and fixed some bugs. I ended up cleaning up the `TextBlock.test.ts` in the backend so the diff probably looks illegible now. For clarity, everything after [75b15bf](https://github.com/iTwin/itwinjs-core/pull/8248/commits/75b15bf6215f4437552e0d4e319d5517c2a7e179) is just whitespace and rearranging. If you want an easier to read diff since you last reviewed, checkout the commit here: [ed98732](https://github.com/iTwin/itwinjs-core/pull/8248/commits/ed9873235249b26815679068b46e99d4096c47a7#diff-0406ba7598de52017fbad055e36dd1b66a8f703fdddea92d0c9712f7a00a41b7)

For reference, this is what the new organization looks like: 

![image](https://github.com/user-attachments/assets/8873874d-5959-431a-9798-7cdb6acbfcea)

MichaelSwigerAtBentley commented on core/common/src/annotation/TextStyle.ts:
```diff
@@ -182,6 +192,7 @@ export class TextStyleSettings {
     this.superScriptOffsetFactor = props.superScriptOffsetFactor ?? defaults.superScriptOffsetFactor;
     this.superScriptScale = props.superScriptScale ?? defaults.superScriptScale;
     this.widthFactor = props.widthFactor ?? defaults.widthFactor;
+    this.tabInterval = undefined === props.tabInterval ? defaults.tabInterval : props.tabInterval;
```
nit: Could this use `??` operator like the pervious values for consistency? 

```suggestion
    this.tabInterval = props.tabInterval ?? defaults.tabInterval;
```

pmconne commented on core/backend/src/annotations/TextBlockLayout.ts:
```diff
@@ -401,6 +412,12 @@ export class RunLayout {
     });
   }
 
+  public applyTabShift(parent: LineLayout, context: LayoutContext) {
```
```suggestion
  /** @internal */
  public applyTabShift(parent: LineLayout, context: LayoutContext) {
```

`LayoutContext` is not exported, no one outside this source file can use it so it doesn't belong in a public method. (I don't know why api-extractor didn't complain about that).

pmconne replied:
Moved it to file scope.

pmconne commented:
Thanks for your patience @claudiaareneee - very nice PR!
